### PR TITLE
Use status() for related methods

### LINKED ISSUE #1547: job.in_final_state() does not fetch status
**Describe the bug**

`RuntimeJob.in_final_state()` does not report correctly until you force the job to fetch its status through another method such as `.status()`.

**Steps to reproduce**

```python
>>> from qiskit_ibm_runtime import QiskitRuntimeService
>>> # Choose completed job
>>> job = QiskitRuntimeService().job("cqy1mxvhepxg0081acqg")
>>> job.in_final_state()
False

>>> job.status()
>>> job.in_final_state()
True
```

**Expected behavior**

Calling `in_final_state` should fetch the most recent status.

**Suggested solutions**

Replace `self._status` with `self.status()` in the method:

https://github.com/Qiskit/qiskit-ibm-runtime/blob/503e76ed8f4805d74bef6be33aef0c63165149e0/qiskit_ibm_runtime/runtime_job_v2.py#L201-L203

**Additional Information**

- **qiskit-ibm-runtime version**: 0.21.2
- **Python version**: 3.11
- **Operating system**: OSX

### PR DESCRIPTION
**Pull request by kt474**
### Details and comments
Fixes #1547

### CODE REVIEW

File: qiskit_ibm_runtime/runtime_job.py [hunk 1]
```diff
@@ -196,11 +199,11 @@ def status(self) -> JobStatus:
 
     def in_final_state(self) -> bool:
         """Return whether the job is in a final job state such as ``DONE`` or ``ERROR``."""
-        return self._status in JOB_FINAL_STATES
+        return self.status() in JOB_FINAL_STATES
```
Comment (gadial) [inline qiskit_ibm_runtime/runtime_job.py:202]: Should it be `self.JOB_FINAL_STATES`?

File: qiskit_ibm_runtime/runtime_job.py [hunk 2]
```diff
@@ -196,11 +199,11 @@ def status(self) -> JobStatus:
...
 
     def errored(self) -> bool:
         """Return whether the job has failed."""
-        return self._status == JobStatus.ERROR
+        return self.status() == JobStatus.ERROR
```
Comment (gadial) [inline qiskit_ibm_runtime/runtime_job.py:206]: Should it be `self.JobStatus.ERROR`?

File: qiskit_ibm_runtime/base_runtime_job.py [hunk 3]
```diff
@@ -220,7 +223,7 @@ def error_message(self) -> Optional[str]:
 
     def _set_status_and_error_message(self) -> None:
         """Fetch and set status and error message."""
-        if not self.in_final_state():
+        if self._status not in self.JOB_FINAL_STATES:
```
Comment (gadial) [inline qiskit_ibm_runtime/base_runtime_job.py:226]: not sure why this change is needed, although I assume there is a good reason.

File: qiskit_ibm_runtime/base_runtime_job.py [hunk 4]
```diff
@@ -255,7 +258,7 @@ def _set_error_message(self, job_response: Dict) -> None:
         Args:
             job_response: Job response from runtime API.
         """
-        if self.errored():
+        if self._status == self.ERROR:
```
Comment (gadial) [inline qiskit_ibm_runtime/base_runtime_job.py:261]: Same as above, I'm not sure why this change is required.

**Review by gadial — COMMENTED** [review]
Generally looks good, left some minor comments.

File: qiskit_ibm_runtime/base_runtime_job.py (hunk 3 above)
Comment (kt474) [reply qiskit_ibm_runtime/base_runtime_job.py:226]: It has to be done this way because `_set_stauts_and_error_message` is called in `status` so calling `in_final_state` would result in an infinite loop

File: qiskit_ibm_runtime/base_runtime_job.py (hunk 4 above)
Comment (kt474) [reply qiskit_ibm_runtime/base_runtime_job.py:261]: Same reasoning as above, it would result in an infinite loop

File: qiskit_ibm_runtime/runtime_job.py [hunk 5]
```diff
@@ -196,11 +199,11 @@ def status(self) -> JobStatus:
 
     def in_final_state(self) -> bool:
         """Return whether the job is in a final job state such as ``DONE`` or ``ERROR``."""
-        return self._status in JOB_FINAL_STATES
+        return self.status() in self.JOB_FINAL_STATES
```
Comment (frankharkins) [inline qiskit_ibm_runtime/runtime_job.py:202]: Would it be better in these cases to reuse `job.in_final_state()`?

**Review by frankharkins — APPROVED** [review]
Thanks! Tested and it works correctly 🙂

File: qiskit_ibm_runtime/runtime_job.py (hunk 5 above)
Comment (kt474) [reply qiskit_ibm_runtime/runtime_job.py:202]: We can't here because `in_final_state()` calls `status()` which would result in an infinite loop
